File: packages/validation-service/src/services/sticker_generator.rs
use qrcode::{types::Color, QrCode};
use uuid::Uuid;

pub struct StickerConfig {
    pub base_url: String,
    pub width: u32,
    pub height: u32,
}

impl Default for StickerConfig {
    fn default() -> Self {
        Self {
            base_url: "https://peek.verse.app".to_string(),
            width: 400,
            height: 480,
        }
    }
}

pub fn generate_sticker_svg(config: &StickerConfig) -> Result<String, anyhow::Error> {
    let community_id = Uuid::new_v4();
    let url = format!("{}/c/{}", config.base_url, community_id);

    let qr_code = QrCode::new(url.as_bytes())?;

    let matrix = qr_code.to_colors();
    let matrix_width = qr_code.width();

    let qr_display_size = (config.width as f32 * 0.65) as u32;

    let quiet_zone_modules = 4;
    let total_modules_with_quiet_zone = matrix_width + (quiet_zone_modules * 2);
    let module_size = qr_display_size as f32 / total_modules_with_quiet_zone as f32;
    let quiet_zone_offset = quiet_zone_modules as f32 * module_size;

    let text_height = 60;
    let total_content_height = qr_display_size + text_height;
    let qr_y = (config.height - total_content_height) / 2;
    let qr_x = (config.width - qr_display_size) / 2;

    let mut qr_modules = String::new();
    for y in 0..matrix_width {
        for x in 0..matrix_width {
            let idx = y * matrix_width + x;
            if matrix[idx] == Color::Dark {
                let px = quiet_zone_offset + (x as f32 * module_size);
                let py = quiet_zone_offset + (y as f32 * module_size);
                qr_modules.push_str(&format!(
                    "    <rect x=\"{:.2}\" y=\"{:.2}\" width=\"{:.2}\" height=\"{:.2}\" rx=\"1\" fill=\"url(#qrGradient)\"/>\n",
                    px, py, module_size, module_size
                ));
            }
        }
    }

    let border_radius = 24;
    let text_y = qr_y + qr_display_size + 50;

    let svg = format!(
        r##"<?xml version="1.0" standalone="yes"?>
<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" viewBox="0 0 {width} {height}">
  <defs>
    <linearGradient id="bgGradient" x1="0%" y1="0%" x2="100%" y2="100%">
      <stop offset="0%" style="stop-color:#FAF5F0;stop-opacity:1" />
      <stop offset="100%" style="stop-color:#FFFFFF;stop-opacity:1" />
    </linearGradient>
    <linearGradient id="qrGradient" x1="0%" y1="0%" x2="100%" y2="100%">
      <stop offset="0%" style="stop-color:#FF6B4A;stop-opacity:1" />
      <stop offset="100%" style="stop-color:#FF8C6B;stop-opacity:1" />
    </linearGradient>
    <filter id="shadow">
      <feDropShadow dx="0" dy="2" stdDeviation="4" flood-opacity="0.1"/>
    </filter>
  </defs>

  <rect width="{width}" height="{height}" rx="{border_radius}" fill="url(#bgGradient)"/>
  <rect x="{qr_x}" y="{qr_y}" width="{qr_display_size}" height="{qr_display_size}"
        rx="12" fill="#FFFFFF" filter="url(#shadow)"/>

  <g transform="translate({qr_x}, {qr_y})">
{qr_modules}  </g>

  <text x="{center_x}" y="{text_y}" text-anchor="middle"
        font-family="system-ui, -apple-system, sans-serif"
        font-size="42" font-weight="700" fill="#2C3E50" letter-spacing="2">PEEK</text>

  <rect x="{line_x}" y="{line_y}" width="80" height="4" rx="2" fill="#4ECDC4"/>
</svg>"##,
        width = config.width,
        height = config.height,
        border_radius = border_radius,
        qr_x = qr_x,
        qr_y = qr_y,
        qr_display_size = qr_display_size,
        qr_modules = qr_modules,
        center_x = config.width / 2,
        text_y = text_y,
        line_x = (config.width - 80) / 2,
        line_y = text_y + 12,
    );

    Ok(svg)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_sticker() {
        let config = StickerConfig::default();
        let result = generate_sticker_svg(&config);
        assert!(result.is_ok());

        let svg = result.unwrap();
        assert!(svg.contains("svg"));
        assert!(svg.contains("PEEK"));
        assert!(svg.contains("/c/"));
    }
}

File: packages/pwa-client/src/services/stickerGenerator.ts
/**
 * Frontend QR Sticker Generator
 *
 * Generates Peek community stickers with QR codes entirely in the browser.
 * No backend needed - works offline in PWA mode.
 */

import QRCode from 'qrcode';

/**
 * Generate a styled Peek sticker SVG with QR code
 *
 * @param communityId Optional community UUID. If not provided, generates a new one.
 * @returns Object with SVG string, community ID, and URL
 */
export async function generateStickerSVG(communityId?: string): Promise<{
  svg: string;
  communityId: string;
  url: string;
}> {
  // Generate UUID if not provided
  const uuid = communityId || crypto.randomUUID();
  const url = `https://peek.verse.app/c/${uuid}`;

  // Generate QR code as SVG (without wrapper)
  const qrSvg = await QRCode.toString(url, {
    type: 'svg',
    errorCorrectionLevel: 'M',
    margin: 2,
    width: 260,
    color: {
      dark: '#2C3E50',
      light: '#FFFFFF',
    },
  });

  // Extract just the path data from qrcode's SVG
  // qrcode library wraps in <svg>, we need just the inner content
  const qrPathMatch = qrSvg.match(/<path[^>]*d="([^"]+)"[^>]*\/>/);
  const qrPath = qrPathMatch ? qrPathMatch[0] : '';

  // Create styled sticker SVG matching backend design
  const styledSvg = `<svg xmlns="http://www.w3.org/2000/svg" width="400" height="480" viewBox="0 0 400 480">
  <defs>
    <!-- Background gradient: cream to white -->
    <linearGradient id="bgGradient" x1="0%" y1="0%" x2="100%" y2="100%">
      <stop offset="0%" style="stop-color:#FAF5F0;stop-opacity:1" />
      <stop offset="100%" style="stop-color:#FFFFFF;stop-opacity:1" />
    </linearGradient>

    <!-- Drop shadow for QR container -->
    <filter id="shadow">
      <feDropShadow dx="0" dy="2" stdDeviation="4" flood-opacity="0.1"/>
    </filter>
  </defs>

  <!-- Background with gradient -->
  <rect width="400" height="480" rx="24" fill="url(#bgGradient)"/>

  <!-- White container for QR code with shadow -->
  <rect x="70" y="60" width="260" height="260" rx="12" fill="#FFFFFF" filter="url(#shadow)"/>

  <!-- QR code -->
  <g transform="translate(70, 60)">
    <svg viewBox="0 0 260 260" width="260" height="260">
      ${qrPath}
    </svg>
  </g>

  <!-- "PEEK" text label -->
  <text
    x="200"
    y="370"
    text-anchor="middle"
    font-family="system-ui, -apple-system, 'Segoe UI', sans-serif"
    font-size="42"
    font-weight="700"
    fill="#2C3E50"
    letter-spacing="2"
  >PEEK</text>

  <!-- Mint accent line -->
  <rect x="160" y="382" width="80" height="4" rx="2" fill="#4ECDC4"/>
</svg>`;

  return {
    svg: styledSvg,
    communityId: uuid,
    url,
  };
}

/**
 * Download sticker as SVG file
 * SVG is preferred for print as it's vector-based (infinite resolution)
 */
export function downloadStickerSVG(svg: string, filename?: string): void {
  const blob = new Blob([svg], { type: 'image/svg+xml' });
  const url = URL.createObjectURL(blob);

  const link = document.createElement('a');
  link.href = url;
  link.download = filename || `peek-sticker-${Date.now()}.svg`;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);

  // Clean up
  URL.revokeObjectURL(url);
}

/**
 * Download sticker as PNG file (rasterized for compatibility)
 *
 * @param svg SVG string to convert
 * @param scale Scale factor for print quality (5 = 2000x2400px at 300 DPI)
 * @param filename Optional filename
 */
export async function downloadStickerPNG(
  svg: string,
  scale: number = 5,
  filename?: string
): Promise<void> {
  return new Promise((resolve, reject) => {
    const canvas = document.createElement('canvas');
    const ctx = canvas.getContext('2d');

    if (!ctx) {
      reject(new Error('Could not get canvas context'));
      return;
    }

    // Set canvas size for print quality
    canvas.width = 400 * scale;  // 2000px at 5x
    canvas.height = 480 * scale; // 2400px at 5x

    // Create image from SVG
    const img = new Image();
    const svgBlob = new Blob([svg], { type: 'image/svg+xml;charset=utf-8' });
    const url = URL.createObjectURL(svgBlob);

    img.onload = () => {
      // Draw scaled image to canvas
      ctx.drawImage(img, 0, 0, canvas.width, canvas.height);

      // Convert to PNG blob
      canvas.toBlob((blob) => {
        if (!blob) {
          reject(new Error('Failed to create PNG blob'));
          return;
        }

        // Download PNG
        const pngUrl = URL.createObjectURL(blob);
        const link = document.createElement('a');
        link.href = pngUrl;
        link.download = filename || `peek-sticker-${Date.now()}.png`;
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);

        // Clean up
        URL.revokeObjectURL(url);
        URL.revokeObjectURL(pngUrl);
        resolve();
      }, 'image/png');
    };

    img.onerror = () => {
      URL.revokeObjectURL(url);
      reject(new Error('Failed to load SVG image'));
    };

    img.src = url;
  });
}

/**
 * Get sticker as data URL for preview/display
 */
export function getStickerDataURL(svg: string): string {
  return `data:image/svg+xml;base64,${btoa(svg)}`;
}
